### PR TITLE
[RTM][ENG-2448] add new landing pages for sso-migration institutions

### PR DESCRIPTION
**Pull request by felliott**
## Purpose

Provide a migration path for users from institutions that no longer support institutional login.

Corresponding [newCAS](https://github.com/CenterForOpenScience/osf-cas/pull/14) & [oldCAS](https://github.com/CenterForOpenScience/cas-overlay/pull/195) tickets create pages that allow the user to request a password reset email if their institution is no longer supported.  This PR adds the necessary osf-side changes to support that.

## Changes

When an existing user from a recently-disabled institution enters OSF, present them with a bespoke page to allow them to set a new, OSF-specific password.  These pages reuse the existing OSF password reset functionality but provide extra context for institutional users in the displayed messages.

## QA Notes

This needs to be tested in combination with either the newCAS or oldCAS PRs.

The expected login workflow is:

* Institutional user clicks "Sign In" button on the OSF frontpage
* They click "Sign in via Institution" button
* Their institution isn't present in the "Select Institution" drop-down, so they click "Can't find my institution"
* They do not have an ORCiD-created account or a previously-OSF-created account.
* They type in their email address into the form.
* They are taken to a page that confirms a "reset password" email has been sent to the provided email.
* If they are the owner of the provided email address, they receive an email that allows them to create an OSF-specific password.

### What are the areas of risk?

Overlooked endpoints. When developing, I assumed that the only access to the "forgot-password" page would be through CAS.  When accessing it directly, I encountered a 500.  That has since been fixed, but could use some critical evaluation

### Any concerns/considerations/questions that development raised?

I don't think so?

## Documentation

This is a very specific endpoint and probably(?) does not need documentation beyond what is provided in the code.

## Side Effects

None expected.

## Ticket

oldCAS: https://openscience.atlassian.net/browse/ENG-2241
newCAS: https://openscience.atlassian.net/browse/ENG-2448

### CODE REVIEW

**Comment by felliott** [comment]
requesting review from @cslzchen

File: framework/auth/views.py [hunk 1]
```diff
@@ -79,6 +78,18 @@ def reset_password_get(auth, uid=None, token=None):
...
+@block_bing_preview
+@collect_auth
+def reset_password_get(auth, uid=None, token=None):
+    return _reset_password_get(auth, uid=uid, token=token)
+
+
+@block_bing_preview
+@collect_auth
+def reset_password_institution_get(auth, uid=None, token=None):
+    return _reset_password_get(auth, uid=uid, token=token)
```
Comment (cslzchen) [inline framework/auth/views.py:90]: There is no need to add an extra layer since the routes definition can handle it by mapping the path to the same view with a different mako. Here is the example:

```python
Rule(
    '/resetpassword-institution/<uid>/<token>/',
    'get',
    auth_views.reset_password_get,
    OsfWebRenderer('public/resetpassword_institution.mako', render_mako_string, trust=False)
),
```

File: framework/auth/views.py [hunk 2]
```diff
@@ -134,8 +145,20 @@ def reset_password_post(uid=None, token=None):
     }
 
 
+def reset_password_institution_post(uid=None, token=None):
+    return reset_password_post(uid=uid, token=token)
```
Comment (cslzchen) [inline framework/auth/views.py:149]: A similar case to `reset_password_institution_get`, there is no need to use a separate view. Here is the updated route definition.

```python
Rule(
    '/resetpassword-institution/<uid>/<token>/',
    'post',
    auth_views.reset_password_post,
    OsfWebRenderer('public/resetpassword_institution.mako', render_mako_string, trust=False)
),
```

File: framework/auth/views.py [hunk 3]
```diff
@@ -134,8 +145,20 @@ def reset_password_post(uid=None, token=None):
...
+
+
+def forgot_password_get():
+    return _forgot_password_get()
+
+
+def forgot_password_institution_get():
+    return _forgot_password_get()
```
Comment (cslzchen) [inline framework/auth/views.py:157]: Unlike the `reset_password_institution_post` and `reset_password_institution_get` that can be merged into `reset_password_post` and `reset_password_get` , this `forgot_password_institution_get()` and `GET /forgotpassword-institution/` route probably needs to be disabled.

This removes confusion in the future so we only have one "entry point", which is implemented in / has been moved to CAS. A better option can be redirecting to the CAS unsupported institution login page.

Here is the route definition to be removed.

```python
Rule(
    '/forgotpassword-institution/',
    'get',
    auth_views.forgot_password_institution_get,
    OsfWebRenderer('public/forgot_password_institution.mako', trust=False)
),
```

Here is what I see if I go to `http://localhost:5000/forgotpassword-institution/`

File: framework/auth/views.py [hunk 4]
```diff
@@ -155,7 +178,7 @@ def forgot_password_get(auth):
     return context
 
 
-def forgot_password_post():
+def forgot_password_post(mail_template=mails.FORGOT_PASSWORD, reset_route='reset_password_get'):
```
Comment (cslzchen) [inline framework/auth/views.py:181]: I see these arguments are defined for both views `forgot_password_post` and `forgot_password_institution_post`. However, they always use the default values without taking advantage of this design. We can combine `forgot_password_post` and `forgot_password_institution_post` since the only differences are are email template and notification message. This not only reduces duplicate code but also makes future maintenance easier.

```python
def forgot_password_post(...)
    return _forgot_password_post(...)

def forgot_password_institution_post(...)
    return _forgot_password_post(...)

def _forgot_password_post(...)
    # everything goes here
```

File: framework/auth/views.py [hunk 5]
```diff
@@ -207,6 +230,69 @@ def forgot_password_post():
     return {}
 
 
+def forgot_password_institution_post(mail_template=mails.FORGOT_PASSWORD_INSTITUTION,
```
Comment (cslzchen) [inline framework/auth/views.py:233]: As mentioned before, merging them into one template and letting arguments control the behavior is a probably the better way to go.

File: website/routes.py [hunk 6]
```diff
@@ -705,6 +705,22 @@ def make_url_map(app):
...
+            OsfWebRenderer('public/resetpassword_institution.mako', render_mako_string, trust=False)
+        ),
+
+        # reset password post - institution
+        Rule(
+            '/resetpassword-institution/<uid>/<token>/',
+            'post',
+            auth_views.reset_password_institution_post,
+            OsfWebRenderer('public/resetpassword_institution.mako', render_mako_string, trust=False)
+        ),
```
Comment (cslzchen) [inline website/routes.py:722]: As mentioned earlier, update both to use the existing view.

File: website/routes.py [hunk 7]
```diff
@@ -789,6 +805,22 @@ def make_url_map(app):
             OsfWebRenderer('public/forgot_password.mako', trust=False)
         ),
 
+        # institutional reset password get
+        Rule(
+            '/forgotpassword-institution/',
+            'get',
+            auth_views.forgot_password_institution_get,
+            OsfWebRenderer('public/forgot_password_institution.mako', trust=False)
+        ),
```
Comment (cslzchen) [inline website/routes.py:814]: As mentioned earlier, remote this routes (thus 404 if hit) or redirect to CAS

File: website/routes.py [hunk 8]
```diff
@@ -789,6 +805,22 @@ def make_url_map(app):
...
+
+        # institutional reset password post
+        Rule(
+            '/forgotpassword-institution/',
+            'post',
+            auth_views.forgot_password_institution_post,
+            OsfWebRenderer('public/forgot_password_institution.mako', trust=False)
+        ),
```
Comment (cslzchen) [inline website/routes.py:822]: This is the one that stays.

**Review by cslzchen — CHANGES_REQUESTED** [review]
Works great locally with both oldCAS and newCAS 🎆. I have a few suggestions listed below. See the comments for details.

- [ ] Use `reset_password_get` and `reset_password_post` for `POST | GET /resetpassword-institution/<uid>/<token>/` since the `mako` template is different.
- [ ] Merge `forgot_password_post` and `forgot_password_institution_views` and use arguments to handle the difference.
- [ ] Remove routes `GET '/forgotpassword-institution/'` to make CAS the only entry point (for users to navigate from) and the only code (for developers to maintain).
- [ ] Add unit tests

File: framework/auth/views.py (hunk 1 above)
Comment (felliott) [reply framework/auth/views.py:90]: D'oh!  I did this in my original implementation, and it didn't work.  It turns out our [`process_rules` function](https://github.com/CenterForOpenScience/osf.io/blob/develop/framework/routing/__init__.py#L159-L169) maps the `Rule`s by building an identifier based off the `view` name and the `renderer` class (not including the args).  The calculated `endpoint` variable for the above `Rule` would be `OsfWebRenderer__forgot_password_get`.   When I used the same view & renderer and only changed the template, one always overwrote the other. That's why I had to break them out into wrapper methods that call the common code.  I thought I had docced this in the commit message, but alas, no.  Sorry about that!  I won't change the code above, but I will amend the commit message to include the reason.  Does that sound good?

File: framework/auth/views.py (hunk 2 above)
Comment (felliott) [reply framework/auth/views.py:149]: Same as above.

File: framework/auth/views.py (hunk 3 above)
Comment (felliott) [reply framework/auth/views.py:157]: Ooh, good catch!  Thanks for that, I will fix it up as suggested.

File: framework/auth/views.py (hunk 4 above)
Comment (felliott) [reply framework/auth/views.py:181]: There is another *small* difference: the non-inst version puts the return message in an `alert`, while the inst version displays it as page block text.  That should be solvable with a new kwarg.  ISTR there was another reason I decided to c-n-p it, rather than reuse a single unified implementation, but I can't for the life of me remember what is was.  I will try reunifying again.  Thank you!

File: website/routes.py (hunk 6 above)
Comment (felliott) [reply website/routes.py:722]: Affected by the `process_rules` overwrite issue, too.

File: website/routes.py (hunk 7 above)
Comment (felliott) [reply website/routes.py:814]: :+1:

File: framework/auth/views.py (hunk 1 above)
Comment (cslzchen) [reply framework/auth/views.py:90]: 👍 Yes, and thanks for the explanation. I thought it might work but didn't try the change out locally.

File: framework/auth/views.py (hunk 4 above)
Comment (cslzchen) [reply framework/auth/views.py:181]: Sounds good! I am also fine to leave them as they are if they can't be merged easily / without adding complexity.

File: framework/auth/views.py [hunk 9]
```diff
@@ -155,9 +187,56 @@ def forgot_password_get(auth):
...
+    :return
+    """
+
+    cas_unsupp_inst_url = cas.get_login_url('', campaign='unsupportedinstitution')
+
+    # if users are logged in, log them out and redirect back to this page
+    if auth.logged_in:
+        return auth_logout(redirect_url=cas_unsupp_inst_url)
+
+    return redirect(cas_unsupp_inst_url)
```
Comment (cslzchen) [inline framework/auth/views.py:208]: One minor issue but not a blocker :)

It is possible that CAS is logged in and OSF is logged out. In this case, CAS will ignore the `campaign=unsupportedinstitution` query param. With `service` as an empty string `''`, this lands user on the generic login success page as shown below.

Screen Shot 2021-02-22 at 4 44 01 PM

Two suggestions:
* Adding OSF home as the default service URL since we expect all usages of CAS with OSF to have one.
* Option 1: force a CAS logout if OSF is already logged out (option 1)

```python
defualt_service_url = web_url_for('index', _absolute=True)  # I think this is for the OSF home.
cas_unsupp_inst_url = cas.get_login_url(defualt_service_url, campaign='unsupportedinstitution')
cas_logout_and_redirect_url = cas.get_logout_url(cas_unsupp_inst_url)

if auth.logged_in:
    return auth_logout(redirect_url=cas_unsupp_inst_url)
return redirect(cas_logout_and_redirect_url)
```

* Option 2: simply skip the `if` condition and always calls OSF `auth_logout()`.

```python
service_url = web_url_for('index', _absolute=True)  # I think this is for the OSF home.
cas_unsupp_inst_url = cas.get_login_url(service_url, campaign='unsupportedinstitution')
return auth_logout(redirect_url=cas_unsupp_inst_url)
```

**Review by cslzchen — APPROVED** [review]
Looks great 🎆 !!!

* 💯 for merging / re-organizing views / templates so the code is more concise with better readability.
* 👍 for adding the unit tests
* 🥇 for the DocStr and comments
* 💊 One minor non-blocking issue for one corner case on the CAS login / logout URL, see the one commit below.

File: framework/auth/views.py (hunk 9 above)
Comment (felliott) [reply framework/auth/views.py:208]: After discussion w/ @cslzchen, we've decided that this is enough of a corner case to not be a blocker for merge.  The user would need to sus the proper endpoint url to trigger this condition, while also maintaining a logged-in CAS & not logged-in OSF.  This is ...difficult, and the reward for overcoming the difficulty is a slightly worse UI.
